File: scripts/download-fide-xml.js
/**
 * download-fide-xml.js
 *
 * Downloads the FIDE standard rating list XML and upserts all rated players
 * into the `players` table by fide_id. Safe to run multiple times.
 *
 * Usage:
 *   DATABASE_URL=... node scripts/download-fide-xml.js
 *   DATABASE_URL=... node scripts/download-fide-xml.js --no-download   # skip download, re-parse existing file
 */

const https = require("https");
const http  = require("http");
const fs    = require("fs");
const path  = require("path");
const readline = require("readline");
const { execSync } = require("child_process");
const { neon } = require("@neondatabase/serverless");

const DB_URL      = process.env.DATABASE_URL;
const DATA_DIR    = path.join(__dirname, "..", "data");
const ZIP_PATH    = path.join(DATA_DIR, "fide_standard.zip");
const EXTRACT_DIR = path.join(DATA_DIR, "fide");
const FIDE_URL    = "https://ratings.fide.com/download/standard_rating_list_xml.zip";
const BATCH_SIZE  = 500;
const NO_DOWNLOAD = process.argv.includes("--no-download");

// ---------------------------------------------------------------------------
// Download with redirect support
// ---------------------------------------------------------------------------
function download(urlStr, dest, redirects = 0) {
  return new Promise((resolve, reject) => {
    if (redirects > 5) return reject(new Error("Too many redirects"));
    const parsed = new URL(urlStr);
    const getter = parsed.protocol === "https:" ? https : http;
    const options = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      headers: { "User-Agent": "Mozilla/5.0 (compatible; chess-distance-etl)" },
    };
    const file = fs.createWriteStream(dest);
    getter.get(options, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        file.close();
        fs.existsSync(dest) && fs.unlinkSync(dest);
        const loc = res.headers.location;
        const next = loc.startsWith("http") ? loc : `${parsed.protocol}//${parsed.hostname}${loc}`;
        return download(next, dest, redirects + 1).then(resolve).catch(reject);
      }
      if (res.statusCode !== 200) {
        file.close();
        return reject(new Error(`HTTP ${res.statusCode}`));
      }
      const total = parseInt(res.headers["content-length"] || "0");
      let received = 0;
      res.on("data", (chunk) => {
        received += chunk.length;
        if (total) {
          const pct = Math.round((received / total) * 100);
          process.stdout.write(`\r  ${pct}% (${(received / 1024 / 1024).toFixed(1)} MB / ${(total / 1024 / 1024).toFixed(1)} MB)`);
        }
      });
      res.pipe(file);
      file.on("finish", () => { file.close(); console.log(""); resolve(); });
      file.on("error", reject);
    }).on("error", reject);
  });
}

// ---------------------------------------------------------------------------
// Unzip (works on Linux with `unzip` or Windows/WSL with PowerShell)
// ---------------------------------------------------------------------------
function extractZip(zipPath, destDir) {
  fs.mkdirSync(destDir, { recursive: true });
  try {
    execSync(`unzip -o "${zipPath}" -d "${destDir}"`, { stdio: "pipe" });
  } catch {
    const zpw = zipPath.replace(/\//g, "\\");
    const ddw = destDir.replace(/\//g, "\\");
    execSync(
      `powershell.exe -Command "Expand-Archive -Path '${zpw}' -DestinationPath '${ddw}' -Force"`,
      { stdio: "pipe", timeout: 120000 }
    );
  }
}

function findXmlFile(dir) {
  const files = fs.readdirSync(dir).filter(f => f.endsWith(".xml"));
  if (!files.length) throw new Error("No XML file found in " + dir);
  // Prefer the standard (not rapid/blitz) file
  const standard = files.find(f => f.toLowerCase().includes("standard")) || files[0];
  return path.join(dir, standard);
}

// ---------------------------------------------------------------------------
// Stream-parse the FIDE XML line by line
// ---------------------------------------------------------------------------
async function parseXml(xmlPath) {
  const rl = readline.createInterface({
    input: fs.createReadStream(xmlPath, { encoding: "utf-8" }),
    crlfDelay: Infinity,
  });

  const players = [];
  let cur = {};
  let inPlayer = false;

  for await (const line of rl) {
    const t = line.trim();
    if (t === "<player>") {
      cur = {};
      inPlayer = true;
    } else if (t === "</player>" && inPlayer) {
      if (cur.fideid && cur.name) players.push(cur);
      if (players.length % 50000 === 0 && players.length > 0) {
        process.stdout.write(`\r  Parsed ${players.length.toLocaleString()} players...`);
      }
      inPlayer = false;
    } else if (inPlayer) {
      // Tags are on their own lines: <tagname>value</tagname>
      const m = t.match(/^<([a-z_]+)>(.*?)<\/\1>$/);
      if (m) cur[m[1]] = m[2];
    }
  }

  process.stdout.write(`\r  Parsed ${players.length.toLocaleString()} players     \n`);
  return players;
}

// ---------------------------------------------------------------------------
// Batch-upsert players into DB
// ---------------------------------------------------------------------------
async function upsertPlayers(players, sql) {
  let done = 0;
  const total = players.length;

  for (let i = 0; i < total; i += BATCH_SIZE) {
    const chunk = players.slice(i, i + BATCH_SIZE);
    const params  = [];
    const clauses = [];

    for (let k = 0; k < chunk.length; k++) {
      const p = chunk[k];
      const fideId = parseInt(p.fideid);
      if (!fideId || !p.name) continue;
      // Prefer OTB title, fall back to women's title
      const title = (p.title && p.title.trim()) || (p.w_title && p.w_title.trim()) || null;
      const base = k * 4;
      clauses.push(`($${base + 1}::integer, $${base + 2}, $${base + 3}, $${base + 4})`);
      params.push(fideId, p.name.trim(), p.country || null, title || null);
    }

    if (!clauses.length) continue;

    await sql.query(
      `INSERT INTO players (fide_id, name, federation, title)
       VALUES ${clauses.join(", ")}
       ON CONFLICT (fide_id) DO UPDATE SET
         name       = EXCLUDED.name,
         federation = EXCLUDED.federation,
         title      = COALESCE(EXCLUDED.title, players.title)`,
      params
    );

    done += chunk.length;
    const pct = Math.round((done / total) * 100);
    process.stdout.write(`\r  Upserted ${done.toLocaleString()} / ${total.toLocaleString()} (${pct}%)`);
  }
  console.log("");
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main() {
  if (!DB_URL) { console.error("No DATABASE_URL"); process.exit(1); }
  const sql = neon(DB_URL);

  fs.mkdirSync(DATA_DIR, { recursive: true });

  // Step 1: Download
  if (NO_DOWNLOAD) {
    console.log("=== Step 1: Skipping download (--no-download) ===");
  } else {
    console.log("=== Step 1: Downloading FIDE standard rating list XML ===");
    if (fs.existsSync(ZIP_PATH)) {
      console.log("  ZIP already exists. Delete data/fide_standard.zip to re-download.");
    } else {
      await download(FIDE_URL, ZIP_PATH);
      console.log(`  Saved to ${ZIP_PATH}`);
    }
  }

  // Step 2: Extract
  console.log("\n=== Step 2: Extracting XML ===");
  extractZip(ZIP_PATH, EXTRACT_DIR);
  const xmlPath = findXmlFile(EXTRACT_DIR);
  console.log(`  ${path.basename(xmlPath)}`);

  // Step 3: Parse
  console.log("\n=== Step 3: Parsing XML ===");
  const players = await parseXml(xmlPath);

  // Step 4: Upsert
  console.log("\n=== Step 4: Upserting into database ===");
  const before = await sql`SELECT count(*) AS n FROM players`;
  await upsertPlayers(players, sql);
  const after = await sql`SELECT count(*) AS n FROM players`;

  console.log(`\n=== Done ===`);
  console.log(`  FIDE players processed: ${players.length.toLocaleString()}`);
  console.log(`  DB players before: ${before[0].n}`);
  console.log(`  DB players after:  ${after[0].n} (+${after[0].n - before[0].n} new)`);
  console.log(`\nNext step: node scripts/batch-pairs.js`);
}

main().catch(console.error);

File: etl/download_fide_xml.py
from __future__ import annotations
"""Download and import the FIDE standard rating list XML.

Downloads the zip, extracts the XML, streams it with iterparse (memory-
efficient even for the ~600 MB unzipped file), then batch-upserts all
~541K rated players into the `players` table by fide_id.

Usage:
    DATABASE_URL=... python etl/download_fide_xml.py
    DATABASE_URL=... python etl/download_fide_xml.py --no-download   # re-parse existing file
"""

import argparse
import sys
import xml.etree.ElementTree as ET
import zipfile
from pathlib import Path

import psycopg.rows
import requests

sys.path.insert(0, str(Path(__file__).parent.parent))
from etl.db import get_conn

DATA_DIR = Path(__file__).parent.parent / "data"
ZIP_PATH = DATA_DIR / "fide_standard.zip"
EXTRACT_DIR = DATA_DIR / "fide"
FIDE_URL = "https://ratings.fide.com/download/standard_rating_list_xml.zip"
BATCH_SIZE = 500
UA = "Mozilla/5.0 (compatible; chess-distance-etl)"


def download_fide_zip() -> None:
    print(f"  Downloading from {FIDE_URL} ...")
    r = requests.get(FIDE_URL, headers={"User-Agent": UA}, stream=True, timeout=300)
    r.raise_for_status()
    total = int(r.headers.get("content-length", 0))
    received = 0
    with open(ZIP_PATH, "wb") as f:
        for chunk in r.iter_content(65536):
            f.write(chunk)
            received += len(chunk)
            if total:
                pct = int(received / total * 100)
                mb = received / 1024 / 1024
                total_mb = total / 1024 / 1024
                print(f"\r  {pct}% ({mb:.1f} MB / {total_mb:.1f} MB)", end="", flush=True)
    print()


def extract_zip() -> Path:
    EXTRACT_DIR.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(ZIP_PATH) as zf:
        zf.extractall(EXTRACT_DIR)
    # Find the standard (not rapid/blitz) XML file
    xmls = list(EXTRACT_DIR.glob("*.xml"))
    if not xmls:
        raise FileNotFoundError(f"No XML file found in {EXTRACT_DIR}")
    standard = next((x for x in xmls if 'standard' in x.name.lower()), xmls[0])
    return standard


def parse_fide_xml(xml_path: Path) -> list[dict]:
    """Stream-parse FIDE XML with iterparse. Memory-efficient for large files."""
    players = []
    for _event, elem in ET.iterparse(xml_path, events=('end',)):
        if elem.tag == 'player':
            fide_id = elem.findtext('fideid')
            name = elem.findtext('name')
            if fide_id and name:
                title = (elem.findtext('title') or '').strip() or \
                        (elem.findtext('w_title') or '').strip() or None
                birthday = elem.findtext('birthday') or None
                birth_year = int(birthday[:4]) if birthday and len(birthday) >= 4 and birthday[:4].isdigit() else None
                players.append({
                    'fide_id': int(fide_id),
                    'name': name.strip(),
                    'federation': (elem.findtext('country') or '').strip() or None,
                    'title': title,
                    'birth_year': birth_year,
                })
            elem.clear()  # free memory
        if len(players) % 50_000 == 0 and players:
            print(f"\r  Parsed {len(players):,} players...", end="", flush=True)
    print(f"\r  Parsed {len(players):,} players     ")
    return players


def upsert_players(conn, players: list[dict]) -> None:
    total = len(players)
    done = 0
    with conn.cursor() as cur:
        for i in range(0, total, BATCH_SIZE):
            chunk = players[i:i + BATCH_SIZE]
            placeholders = ','.join(['(%s,%s,%s,%s,%s)'] * len(chunk))
            params: list = []
            for p in chunk:
                params.extend([p['fide_id'], p['name'], p['federation'], p['title'], p['birth_year']])
            cur.execute(f"""
                INSERT INTO players (fide_id, name, federation, title, birth_year)
                VALUES {placeholders}
                ON CONFLICT (fide_id) DO UPDATE SET
                    name       = EXCLUDED.name,
                    federation = EXCLUDED.federation,
                    title      = EXCLUDED.title,
                    birth_year = COALESCE(players.birth_year, EXCLUDED.birth_year)
            """, params)
            conn.commit()
            done += len(chunk)
            pct = int(done / total * 100)
            print(f"\r  Upserted {done:,} / {total:,} ({pct}%)", end="", flush=True)
    print()


def main() -> None:
    parser = argparse.ArgumentParser(description="Download and import FIDE rating list XML")
    parser.add_argument("--no-download", action="store_true", help="Skip download, re-parse existing file")
    args = parser.parse_args()

    DATA_DIR.mkdir(parents=True, exist_ok=True)
    conn = get_conn()

    # Step 1: Download
    if args.no_download:
        print("=== Step 1: Skipping download (--no-download) ===")
    else:
        print("=== Step 1: Downloading FIDE standard rating list XML ===")
        if ZIP_PATH.exists():
            print(f"  ZIP already exists at {ZIP_PATH}. Delete it to re-download.")
        else:
            download_fide_zip()
            print(f"  Saved to {ZIP_PATH}")

    # Step 2: Extract
    print("\n=== Step 2: Extracting XML ===")
    xml_path = extract_zip()
    print(f"  {xml_path.name}")

    # Step 3: Parse
    print("\n=== Step 3: Parsing XML ===")
    players = parse_fide_xml(xml_path)

    # Step 4: Upsert
    print("\n=== Step 4: Upserting into database ===")
    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        cur.execute("SELECT count(*) AS n FROM players")
        before = cur.fetchone()['n']

    upsert_players(conn, players)

    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        cur.execute("SELECT count(*) AS n FROM players")
        after = cur.fetchone()['n']

    print(f"\n=== Done ===")
    print(f"  FIDE players processed: {len(players):,}")
    print(f"  DB players before:      {before:,}")
    print(f"  DB players after:       {after:,} (+{after - before:,} new)")
    print(f"\nNext step: python etl/batch_pairs.py")
    conn.close()


if __name__ == "__main__":
    main()
